### PR TITLE
Smp grad accum

### PR DESCRIPTION
**Pull request by sgugger**
# What does this PR do?

This PR adds support for gradient accumulation in `SageMakerTrainer`. It has been tested on the glue script with success (with and without gradient accumulation passed along).

### CODE REVIEW

File: src/transformers/sagemaker/trainer_sm.py [hunk 1]
```diff
@@ -108,7 +107,7 @@ def _wrap_model(self, model, training=True):
             # Wrapping the base model twice in a DistributedModel will raise an error.
             if isinstance(self.model_wrapped, smp.model.DistributedModel):
                 return self.model_wrapped
-            return smp.DistributedModel(model)
+            return smp.DistributedModel(model, backward_passes_per_step=self.args.gradient_accumulation_steps)
```
Comment (sgugger) [inline src/transformers/sagemaker/trainer_sm.py:110]: This does the equivalent of `no_sync` in regular DDP.

File: src/transformers/training_args.py [hunk 2]
```diff
@@ -737,6 +737,13 @@ def place_model_on_device(self):
         """
         return True
 
+    @property
+    def _no_sync_in_gradient_accumulation(self):
+        """
+        Whether or not to use no_sync for the gradients when doing gradient accumulation.
+        """
+        return not self.deepspeed
```
Comment (sgugger) [inline src/transformers/training_args.py:745]: This is introduced to make it easy to skip the no_sync part in subclasses.

**Review by anirudh2290 — APPROVED** [review]
Thanks a lot for the fix! LGTM.

**Review by LysandreJik — APPROVED** [review]
Very nice! LGTM!

**Pull request merged**